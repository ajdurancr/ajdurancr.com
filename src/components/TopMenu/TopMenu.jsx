import React, { useState, useEffect } from 'react'
import propTypes from 'prop-types'
import classnames from 'classnames'

import buildSiteMenu from '../../helpers/buildSiteMenu'
import { getParentPageUrl } from '../../helpers/url'

import SubMenu from './SubMenu'
import styles from './TopMenu.styles.sass'

const TopMenu = ({ menu, currentPageUrl }) => {
  const [parentPageUrl, setParentPageUrl] = useState(getParentPageUrl(currentPageUrl))
  const [menuItems, setMenuItems] = useState(buildSiteMenu(menu))

  useEffect(() => {
    setMenuItems(buildSiteMenu(menu))
  }, [menu])

  useEffect(() => {
    setParentPageUrl(getParentPageUrl(currentPageUrl))
  }, [currentPageUrl])

  return (
    <nav className={styles.menu}>
      <ul>
        {menuItems.map(({ name, url, submenu }) => (
          <li className={classnames({ [styles.activeItem]: parentPageUrl === url })} key={url}>
            <a href={url}>{name}</a>
            {submenu && <SubMenu submenu={submenu} />}
          </li>
        ))}
      </ul>
    </nav>
  )
}

TopMenu.defaultProps = {
  menu: {},
}

TopMenu.propTypes = {
  currentPageUrl: propTypes.string.isRequired,
  menu: propTypes.shape({
    name: propTypes.string,
    submenu: propTypes.arrayOf(
      propTypes.shape({
        alias: propTypes.string,
        name: propTypes.string,
        page: propTypes.object,
        submenu: propTypes.arrayOf(propTypes.object),
        url: propTypes.string,
      }),
    ),
  }),
}

export default TopMenu
