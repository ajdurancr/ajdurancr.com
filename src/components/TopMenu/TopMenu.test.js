import React from 'react'
import { shallow } from 'enzyme'

import TopMenu from './TopMenu'
import topMenu from '../../mocks/contentful/topMenu'

const currentPageUrl = '/'

describe('TopMenu', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<TopMenu currentPageUrl={currentPageUrl} menu={topMenu.fields} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('renders correctly with default props', () => {
    const wrapper = shallow(<TopMenu currentPageUrl={currentPageUrl} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('renders correctly when submenu item selected', () => {
    const wrapper = shallow(<TopMenu currentPageUrl="/resume/education" menu={topMenu.fields} />)

    expect(wrapper).toMatchSnapshot()
  })
})
