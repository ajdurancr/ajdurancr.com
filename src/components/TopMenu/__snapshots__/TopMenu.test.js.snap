// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopMenu renders correctly 1`] = `
<nav
  className="menu"
>
  <ul>
    <li
      className="activeItem"
      key="/"
    >
      <a
        href="/"
      >
        Home
      </a>
    </li>
    <li
      className=""
      key="/blog"
    >
      <a
        href="/blog"
      >
        Blog
      </a>
    </li>
    <li
      className=""
      key="/resume"
    >
      <a
        href="/resume"
      >
        Resume
      </a>
      <SubMenu
        submenu={
          Array [
            Object {
              "name": "Education",
              "submenu": undefined,
              "url": "/resume/education",
            },
            Object {
              "name": "Work Experience",
              "submenu": undefined,
              "url": "/resume/work-experience",
            },
          ]
        }
      />
    </li>
    <li
      className=""
      key="/portfolio"
    >
      <a
        href="/portfolio"
      >
        Portfolio
      </a>
    </li>
    <li
      className=""
      key="/about"
    >
      <a
        href="/about"
      >
        About
      </a>
    </li>
    <li
      className=""
      key="http://www.google.com"
    >
      <a
        href="http://www.google.com"
      >
        Menu - DCR.IO External link
      </a>
    </li>
    <li
      className=""
      key="http://www.google.com"
    >
      <a
        href="http://www.google.com"
      >
        Google
      </a>
    </li>
  </ul>
</nav>
`;

exports[`TopMenu renders correctly when submenu item selected 1`] = `
<nav
  className="menu"
>
  <ul>
    <li
      className=""
      key="/"
    >
      <a
        href="/"
      >
        Home
      </a>
    </li>
    <li
      className=""
      key="/blog"
    >
      <a
        href="/blog"
      >
        Blog
      </a>
    </li>
    <li
      className="activeItem"
      key="/resume"
    >
      <a
        href="/resume"
      >
        Resume
      </a>
      <SubMenu
        submenu={
          Array [
            Object {
              "name": "Education",
              "submenu": undefined,
              "url": "/resume/education",
            },
            Object {
              "name": "Work Experience",
              "submenu": undefined,
              "url": "/resume/work-experience",
            },
          ]
        }
      />
    </li>
    <li
      className=""
      key="/portfolio"
    >
      <a
        href="/portfolio"
      >
        Portfolio
      </a>
    </li>
    <li
      className=""
      key="/about"
    >
      <a
        href="/about"
      >
        About
      </a>
    </li>
    <li
      className=""
      key="http://www.google.com"
    >
      <a
        href="http://www.google.com"
      >
        Menu - DCR.IO External link
      </a>
    </li>
    <li
      className=""
      key="http://www.google.com"
    >
      <a
        href="http://www.google.com"
      >
        Google
      </a>
    </li>
  </ul>
</nav>
`;

exports[`TopMenu renders correctly with default props 1`] = `
<nav
  className="menu"
>
  <ul />
</nav>
`;
